perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -1734,11 +1737,140 @@ protected boolean hasCandidateExistingMember(Candidate candidate, RichMember ric
 
+	/**
+	 * Sync resources from groupStructureResources attribute to the group
+	 * tree with the root in the given base group. If some resources are not found
+	 * or the assignment failed, there is a new message added to the skippedMessages list.
+	 *
+	 * @param sess perun session
+	 * @param baseGroup group structure sync base group
+	 * @param skippedMessages list where are added messages about skipped operations
+	 */
+	private void syncResourcesForSynchronization(PerunSession sess, Group baseGroup, AttributeDefinition loginAttr,
+	                                             List<String> skippedMessages) {
+		Attribute syncedResourcesAttr;
+
+		try {
+			syncedResourcesAttr = perunBl.getAttributesManagerBl()
+					.getAttribute(sess, baseGroup, A_G_D_GROUP_STRUCTURE_RESOURCES);
+		} catch (WrongAttributeAssignmentException | AttributeNotExistsException e) {
+			throw new InternalErrorException("Failed to obtain the groupStructureResources attribute for structure synchronization.", e);
+		}
+
+		if (syncedResourcesAttr.getValue() == null) {
+			// no resources should be synced
+			return;
+		}
+
+		// load all subgroups with logins once, so it can be reused in other methods
+		Map<String, Group> groupsByLogins = getAllSubGroupsWithLogins(sess, baseGroup, loginAttr);
+
+		syncedResourcesAttr.valueAsMap().forEach((resourceId, groupLogins) ->
+			syncResourceInStructure(sess, resourceId, groupLogins, baseGroup, groupsByLogins, skippedMessages));
+	}
+
+	/**
+	 * Assign resource with given id to groups with logins defined in parameter 'rawGroupLogins'.
+	 *
+	 * If the rawGroupLogins value is empty, then the resource is assigned to the whole structure.
+	 * The rawGroupLogins are in format: 'login1,login2,login3,'. If the login contains a '\' or ','
+	 * characters, they must be escaped by the '\' character.
+	 *
+	 * @param sess perun session
+	 * @param rawResourceId id of a resource which should be set
+	 * @param rawGroupLogins group login separated with a comma ','
+	 * @param baseGroup base group which is used if the rawGroupLogins is empty
+	 * @param groupsByLogins map containing groups with by their logins
+	 * @param skippedMessages a list with skipped messages, other messages are added to this list
+	 */
+	private void syncResourceInStructure(PerunSession sess, String rawResourceId, String rawGroupLogins,
+	                                     Group baseGroup, Map<String, Group> groupsByLogins,
+	                                     List<String> skippedMessages) {
+		int resourceId = Integer.parseInt(rawResourceId);
+		try {
+			Resource resource = perunBl.getResourcesManagerBl().getResourceById(sess, resourceId);
+			// if no group logins are specified, assign resource to the whole tree
+			if (rawGroupLogins.isEmpty()) {
+				assignResourceToGroupAndSubGroups(sess, baseGroup, resource, skippedMessages);
+			} else {
+				List<String> groupLogins = BeansUtils.parseEscapedListValue(rawGroupLogins);
+				groupLogins.forEach(login ->
```

Comment:
I think you do not want to iterate over each login and then for each iteration call method syncResourceInStructure with all logins again.

Reply:
Yes, you are right. I have fixed it.